File: fsharpfunctions/DataProcessing/IssueDataV2.fs
namespace FSharpFunctions
open System
open System.Collections.Generic


//new types
type SubService = {Name: string; Count: int}
type Service = {Name: string; Count: int; SubServiceCounts: option<list<SubService>>}

type IssueDataV2 = {
    Id: string
    Source: Source
    EntryType: EntryType //we know what possible values are
    Timestamp: DateTime
    TotalOpenIssues: int
    MissingTags: int
    CountByPriority: option<Map<string,int>> //immutable key-value store
    CountByService: option<list<Service>>  // union type allow both old and new data types.
}

File: csharpfunctions/DataProcessing/IssueDataV2.cs
using csharpfunctions.DataProcessing;
using System;
using System.Collections.Generic;
using System.Text;

namespace csharpfunctions
{
    public class IssueDataV2
    {
        public string Id { get; set; }
        public string Source { get; set; }
        public string EntryType { get; set; }
        public DateTime Timestamp { get; set; }
        public int TotalOpenIssues { get; set; }
        public int MissingTags { get; set; }
        public Dictionary<string, int> CountByPriority { get; set; }
        public List<ItemCount> CountByService { get; set; }
    }
}
